---
title: "Single_Cell_ATAC"
author: "Alos Diallo"
date: "2022-10-31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown


```{r cars, include=FALSE}
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(EnsDb.Hsapiens.v75)
library(ggplot2)
library(patchwork)
library(hdf5r)
library(biovizBase)
set.seed(1234)
```



```{r pressure, echo=FALSE}
counts <- Read10X_h5(filename = "/Users/adiallo/Desktop/Dartmouth/Christiansen_lab/CF_Project/scATAC/all/outs/filtered_peak_bc_matrix.h5")
metadata <- read.csv(
  file = "/Users/adiallo/Desktop/Dartmouth/Christiansen_lab/CF_Project/scATAC/all/outs/singlecell.csv",
  header = TRUE,
  row.names = 1
)

chrom_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  genome = 'hg19',
  fragments = '/Users/adiallo/Desktop/Dartmouth/Christiansen_lab/CF_Project/scATAC/all/outs/fragments.tsv.gz',
  min.cells = 10,
  min.features = 200
)

pbmc <- CreateSeuratObject(
  counts = chrom_assay,
  assay = "peaks",
  meta.data = metadata
)


```
```{r}
pbmc
pbmc[['peaks']]

granges(pbmc)
```
```{r warning=FALSE}
# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)

# change to UCSC style since the data was mapped to hg19
seqlevelsStyle(annotations) <- 'UCSC'

# add the gene information to the object
Annotation(pbmc) <- annotations
```

```{r}
# compute nucleosome signal score per cell
pbmc <- NucleosomeSignal(object = pbmc)

# compute TSS enrichment score per cell
pbmc <- TSSEnrichment(object = pbmc, fast = FALSE)

# add blacklist ratio and fraction of reads in peaks
pbmc$pct_reads_in_peaks <- pbmc$peak_region_fragments / pbmc$passed_filters * 100
pbmc$blacklist_ratio <- pbmc$blacklist_region_fragments / pbmc$peak_region_fragments



```
```{r}
pbmc$high.tss <- ifelse(pbmc$TSS.enrichment > 2, 'High', 'Low')
TSSPlot(pbmc, group.by = 'high.tss') + NoLegend()

pbmc$nucleosome_group <- ifelse(pbmc$nucleosome_signal > 4, 'NS > 4', 'NS < 4')
FragmentHistogram(object = pbmc, group.by = 'nucleosome_group')
```
```{r}
VlnPlot(
  object = pbmc,
  features = c('pct_reads_in_peaks', 'peak_region_fragments',
               'TSS.enrichment', 'blacklist_ratio', 'nucleosome_signal'),
  pt.size = 0.1,
  ncol = 5
)
```
```{r}
pbmc <- subset(
  x = pbmc,
  subset = peak_region_fragments > 3000 &
    peak_region_fragments < 20000 &
    pct_reads_in_peaks > 15 &
    blacklist_ratio < 0.05 &
    nucleosome_signal < 4 &
    TSS.enrichment > 2
)
pbmc
```

```{r}
pbmc <- RunTFIDF(pbmc)
pbmc <- FindTopFeatures(pbmc, min.cutoff = 'q0')
pbmc <- RunSVD(pbmc)
```
```{r}
DepthCor(pbmc)
```

```{r}
pbmc <- RunUMAP(object = pbmc, reduction = 'lsi', dims = 2:30)
pbmc <- FindNeighbors(object = pbmc, reduction = 'lsi', dims = 2:30)
pbmc <- FindClusters(object = pbmc, verbose = FALSE, algorithm = 3)
DimPlot(object = pbmc, label = TRUE) + NoLegend()
```
```{r}
gene.activities <- GeneActivity(pbmc)
```
```{r}
# add the gene activity matrix to the Seurat object as a new assay and normalize it
pbmc[['RNA']] <- CreateAssayObject(counts = gene.activities)
pbmc <- NormalizeData(
  object = pbmc,
  assay = 'RNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(pbmc$nCount_RNA)
)
```
```{r}
DefaultAssay(pbmc) <- 'RNA'

FeaturePlot(
  object = pbmc,
  features = c('CD14', 'FCGR3B'),
  pt.size = 0.1,
  max.cutoff = 'q95',
  ncol = 3
)
```
```{r}
pbmc_rna <- readRDS("/Users/adiallo/Desktop/Dartmouth/Christiansen_lab/CF_Project/scATAC/all/outs/pbmc_10k_v3.rds")

transfer.anchors <- FindTransferAnchors(
  reference = pbmc_rna,
  query = pbmc,
  reduction = 'cca'
)

predicted.labels <- TransferData(
  anchorset = transfer.anchors,
  refdata = pbmc_rna$celltype,
  weight.reduction = pbmc[['lsi']],
  dims = 2:30
)

pbmc <- AddMetaData(object = pbmc, metadata = predicted.labels)
```

```{r}
plot1 <- DimPlot(
  object = pbmc_rna,
  group.by = 'celltype',
  label = TRUE,
  repel = TRUE) + NoLegend() + ggtitle('scRNA-seq')

plot2 <- DimPlot(
  object = pbmc,
  group.by = 'predicted.id',
  label = TRUE,
  repel = TRUE) + NoLegend() + ggtitle('scATAC-seq')

plot1 + plot2
```
```{r}
pbmc <- subset(pbmc, idents = 14, invert = TRUE)
pbmc <- RenameIdents(
  object = pbmc,
  '0' = 'B cell progenitor',
  '1' = 'CD14+ Monocytes',
  '2' = 'CD16+ Monocytes',
  '3' = 'CD4 Memory',
  '4' = 'CD4 Naive',
  '5' = 'CD8 effector',
  '6' = 'CD8 Naive',
  '7' = 'Dendridic ell',
  '8' = 'Double negative T cell',
  '9' = 'NK bright',
  '10' = 'NK dim',
  '11' = 'DC',
  '12' = 'NK CD56bright',
  '13' = 'pDC'
)
```



```{r}

cov_plot <- CoveragePlot(
  object = pbmc,
  region = "chr5-140631856-140632980",
  annotation = FALSE,
  peaks = FALSE
)
cov_plot


CoveragePlot(
  object = pbmc,
  region = c("CD14",'FCGR3B'),
  peaks = FALSE
)

```

